declarativewidgets/proposal.md

```diff
@@ -0,0 +1,68 @@
+## Subproject goals, scope and functionality
+
+### Concept 
+
+>What if the ecosystem around Project Jupyter included a pervasive set of rich notebook user interfaces (a.k.a. widgets) using declarative markup that can interact with other code authored in the notebook? 
+
+A very powerful feature of Jupyter Notebooks is the ability to mix interactive widgets along with your code and data. These widgets turn the notebook into more of a web application, allowing the author to create a User Interface (UI) to the code entered in the Notebook’s cells. Rather than editing code and executing cells, the consumer of the Notebook can interact with the code by moving sliders, filling a form, pressing a button, etc.
+
+This incubation proposal pertains to the creation of a Jupyter extension that would allow developers to create a compendium of language agnostic widgets that can support the Jupyter ecosystem of language backends. The extension would leverage web-components and Polymer in combination with the existing IPywidgets libraries.
+
+### Goals
+* Give the author access to a vast toolbox of components and a simple yet powerful authoring experience.
+	* Enable quick and easy integration between the  ecosystem of Web Components (e.g., widgets in the Polymer catalog) and the Jupyter Interactive Notebook.
+	* Provide Polymer based binding support to allow widgets to be aware of interactivity within the scope of the notebook namespace. 
+* Minimize the effort to fully support interactive widgets on all Notebook supported languages. 
+	* Instantiation of widgets within the notebook using HTML. 
+	* Achieve compatibility and portability between the ecosystem of widgets and the Jupyter Interactive Notebook, regardless of the language used in the Notebook (i.e. Python, R, Scala, etc.).  
+
+### Scope
+The objective of this incubation project is to establish a baseline prototype that begins to blur the lines between a notebook and an application while bootstrapping the ecosystem of notebook compatible widgets. It will build on top of the foundation set by IPywidgets.
+
+This incubation effort can be scoped as:
+
+1. Identification of the minimum essential pieces of IPyWidgets that need to exist in all language kernels
+2. Building a portable layer can leverage an existing and ever growing ecosystem of web widgets.
+3. Providing a widget framework that works across the various and growing set of languages that can be hosted in a Jupyter Interactive Notebook. 
+
+### Functionality
+
+As outlined in this [blog post](http://blog.ibmjstart.net/2015/08/21/declarative-widget-system-for-jupyter-notebooks/), the intent of this proposal is to provide support for:
+
+* A **function element** which will allow the running of a segment of code within the notebook as a side effect of a user's interactivity with UI elements. By using a function defined in a code cell, the author can connect other visual elements to set the function arguments, invoke it and display its result.
```

Comment:
My concern is that exposing code segments directly in the frontend will introduce language specific aspects. The alternative to the manage this through synchronized state that has clear JSON expressions. I think exploring the state versus code aspects of this will be one of the interesting question during incubation.